### ISSUE TITLE
System.ArgumentException when triing SpaServices.Extensions

### ISSUE THREAD

**Issue by JanEggers** [aspnet/JavaScriptServices#1469]
Im getting the following exception if a client connects using your new amazing extensions:

```
app.UseSpa((o) =>
            {
                if (env.IsDevelopment()) {
                    o.UseProxyToSpaDevelopmentServer("http://localhost:4200");
                }
            });
```

also hmr is quite slow but that may be related to the error:

System.ArgumentException: Der 'User-Agent'-Header muss mit der entsprechenden Eigenschaft oder Methode geändert werden.
Parametername: name
   bei System.Net.WebHeaderCollection.ThrowOnRestrictedHeader(String headerName)
   bei System.Net.WebHeaderCollection.Set(String name, String value)
   bei Microsoft.AspNetCore.SpaServices.Extensions.Proxy.SpaProxy.<AcceptProxyWebSocketRequest>d__8.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei Microsoft.AspNetCore.SpaServices.Extensions.Proxy.SpaProxy.<PerformProxyRequest>d__4.MoveNext()

here are my nuget references:

<PackageReference Include="Microsoft.AspNetCore.Mvc.Cors" Version="2.0.1" />

    <PackageReference Include="Microsoft.AspNetCore.Mvc.Formatters.Json" Version="2.0.1" />

    <PackageReference Include="Microsoft.AspNetCore.Server.IISIntegration" Version="2.0.1" />
    <PackageReference Include="Microsoft.AspNetCore.Server.Kestrel" Version="2.0.1" />
    <PackageReference Include="Microsoft.AspNetCore.SpaServices.Extensions" Version="2.0.0-rc1-final" />
    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.0.1" />
    <PackageReference Include="Microsoft.AspNetCore.ResponseCompression" Version="2.0.1" />
    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.0.0" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="2.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Configuration" Version="2.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="2.0.0" />
    <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="2.0.0" />
    <PackageReference Include="Serilog.Extensions.Logging" Version="2.0.2" />
    <PackageReference Include="Serilog.Settings.Configuration" Version="2.4.0" />
    <PackageReference Include="Serilog.Sinks.Async" Version="1.1.0" />
    <PackageReference Include="Serilog.Sinks.RollingFile" Version="3.3.0" />

**Comment by JanEggers** [comment]
here is the line:

https://github.com/aspnet/JavaScriptServices/blob/c3964a0437a12166d11ead75e1ff35eefc0181bd/src/Microsoft.AspNetCore.SpaServices.Extensions/Proxying/SpaProxy.cs#L207

that is the problem:

https://connect.microsoft.com/VisualStudio/feedback/details/840553/system-net-websockets-clientwebsocket-does-not-allow-to-set-the-http-header-user-agent

and here is a workaround:

https://stackoverflow.com/questions/15894844/how-can-i-set-user-agent-and-referer-headers-when-using-clientwebsocket-in-net

**Comment by SteveSandersonMS** [comment]
Thanks for reporting this. What I don't quite understand is why you're getting this, but nobody else ever seems to have done (and I know I haven't).

Do you have repro steps? If you can provide a minimal code sample that triggers this issue I'm sure we can fix it.

**Comment by JanEggers** [comment]
hi Steve,

i have reproSteps and a repostitory:

https://github.com/JanEggers/SpaExtensionsRepro

dotnet new angular

switch targetframework to net461

replace "all package" by minimunm required packages to get stuff running

F5

**Comment by JanEggers** [comment]
hey @SteveSandersonMS any news on this one or do you need more info?

**Comment by mjrousos** [comment]
I'm hitting this as well. Any updates on a fix?

**Comment by mjrousos** [comment]
I figured out why this doesn't occur more often - it appears to only be a problem when targetting .NET Framework (as opposed to .NET Core). @SteveSandersonMS, to repro, take any "hello world" ASP.NET Core SPA app and update the csproj to have `<TargetFramework>net462</TargetFramework>` instead of `<TargetFramework>netcoreapp2.0</TargetFramework>` (you'll also need to depend on Microsoft.AspNetCore and Microsoft.AspNetCore.Mvc instead of Microsoft.AspNetCore.All) and you should see this exception. If it's helpful, I can push my sample to GitHub somewhere.

It's not clear to me yet why this only happens on NetFX and not on NetCore (probably some System.Net difference), but in the short-term, I have a simple work-around in case it helps you, @JanEggers.

In your Startup.cs file, in the `Configure` method, add custom middleware like this inside of the `if (env.IsDevelopment)` block (so that it only applies when running in development environments:

```CSharp
// Work around https://github.com/aspnet/JavaScriptServices/issues/1469
app.Use((httpContext, next) =>
{
    if (httpContext.Request.Path.HasValue &&
        httpContext.Request.Path.Value.Contains("sockjs-node") &&
        httpContext.Request.Path.Value.Contains("websocket"))
    {
        if (httpContext.Request.Headers.ContainsKey("User-Agent"))
        {
            httpContext.Request.Headers.Remove("User-Agent");
        }
    }
    return next();
});
```

The failure occurs because `SpaProxy` is attempting to copy a User-Agent header from an incoming request to its corresponding outgoing one but `WebHeaderCollection` doesn't allow User-Agent headers to be set in that way. This middleware, just removes the User-Agent header from incoming requests that include both 'sockjs-node' and 'websocket' in their request path.

**Comment by JanEggers** [comment]
@mjrousos  thx for the workaround, fortunatly im able to mock my backend so i just use ng serve without aspnet for now. I dont think you also need to provide a repro as i already did one.

**Comment by SteveSandersonMS** [comment]
Thanks for all the info here. It's now fixed in the repo. The fix will be included in the 2.0.0 RC2 package, hopefully today.

**Comment by mjrousos** [comment]
Great, thanks!
